File: selection/_019/Main.java
package selection._019;

import java.util.Arrays;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int d = sc.nextInt();
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[] shopPoints = new int[n + 1];
        int[] orderPoints = new int[m];

        shopPoints[n] = d;
        for (int i = 1; i < n; i++) {
            shopPoints[i] = sc.nextInt();
        }
        for (int i = 0; i < m; i++) {
            orderPoints[i] = sc.nextInt();
        }

        sc.close();

        Arrays.sort(shopPoints);

        int result = 0;
        for (int i = 0; i < m; i++) {
            int lb = lower_bound(shopPoints, orderPoints[i]);
            int dist = Math.abs(shopPoints[lb] - orderPoints[i]);
            if (lb - 1 >= 0) {
                dist = Math.min(dist, Math.abs(shopPoints[lb - 1] - orderPoints[i]));
            }
            result += dist;
        }

        System.out.println(result);
    }

    static int lower_bound(int[] a, int key) {
        int l = -1, r = a.length;
        // 条件を満たすかどうかの堺目が分かる状態になるのはl + 1 = rの状態
        while (r - l > 1) {
            int mid = (l + r) / 2;
            if (a[mid] >= key) {
                r = mid;
            } else {
                l = mid;
            }
        }

        return r;
    }
}

File: selection/_019/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func main() {
	var d, n, m int
	fmt.Scan(&d, &n, &m)

	s := make([]int, n+1)
	for i := 1; i < n; i++ {
		fmt.Scan(&s[i])
	}
	s[n] = d

	t := make([]int, m)
	for i := 0; i < m; i++ {
		fmt.Scan(&t[i])
	}

	sort.Ints(s)

	result := 0
	for i := 0; i < m; i++ {
		it := lower_bound(s, t[i])
		tmp := math.Abs(float64(s[it] - t[i]))
		if it > 0 {
			tmp = math.Min(tmp, math.Abs(float64(s[it-1]-t[i])))
		}
		result += int(tmp)
	}

	fmt.Println(result)
}

func lower_bound(a []int, t int) int {
	ng, ok := -1, len(a)

	for ok-ng > 1 {
		mid := (ok + ng) / 2

		if a[mid] >= t {
			ok = mid
		} else {
			ng = mid
		}
	}

	return ok
}
